lib/shared/files.js

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2018 resin.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict'
+
+const _ = require('lodash')
+const path = require('path')
+const Bluebird = require('bluebird')
+const fs = Bluebird.promisifyAll(require('fs'))
+
+/* eslint-disable lodash/prefer-lodash-method */
+
+/**
+ * @summary Get file metadata
+ * @function
+ * @private
+ *
+ * @param {String} dirname - directory name
+ * @param {String} [basename] - custom basename to append
+ * @returns {Object} file metadata
+ *
+ * @example
+ * try {
+ *   const file = files.getFileMetadataSync('/home/user')
+ *   console.log(`Is ${file.basename} a directory? ${file.isDirectory}`)
+ * } catch (error) {
+ *   console.error(error)
+ * }
+ */
+exports.getFileMetadataSync = (dirname, basename = '') => {
+  // TODO(Shou): use path.parse object information here
+  const fullpath = path.join(dirname, basename)
+  const pathObj = path.parse(fullpath)
+
+  // TODO(Shou): this is not true for Windows, figure out Windows hidden files
+  const isHidden = pathObj.base.startsWith('.')
+  const stats = fs.statSync(fullpath)
+
+  return {
+    basename: pathObj.base,
+    dirname: pathObj.dir,
+    fullpath,
+    extension: pathObj.ext.replace('.', ''),
+    name: pathObj.name,
+    isDirectory: stats.isDirectory(),
```

Comment:
Maybe worth having the `stats` object on the metadata in here? Would allow for easier future extensions; i.e. `metadata.stats.isDirectory()` (same for the async method obv.)

Reply:
I would do that but at least `isDirectory` needs to be a value because of filters so either way we need to do some custom handling of it.